In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv('C:/Users/Saad/Downloads/archive/Pakistan Largest Ecommerce Dataset.csv', encoding='latin-1', low_memory=False)

In [4]:
df.shape

(1048575, 26)

In [5]:
df.dtypes


item_id                  float64
status                    object
created_at                object
sku                       object
price                    float64
qty_ordered              float64
grand_total              float64
increment_id              object
category_name_1           object
sales_commission_code     object
discount_amount          float64
payment_method            object
Working Date              object
BI Status                 object
 MV                       object
Year                     float64
Month                    float64
Customer Since            object
M-Y                       object
FY                        object
Customer ID              float64
Unnamed: 21              float64
Unnamed: 22              float64
Unnamed: 23              float64
Unnamed: 24              float64
Unnamed: 25              float64
dtype: object

In [6]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 26 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   item_id                584524 non-null  float64
 1   status                 584509 non-null  object 
 2   created_at             584524 non-null  object 
 3   sku                    584504 non-null  object 
 4   price                  584524 non-null  float64
 5   qty_ordered            584524 non-null  float64
 6   grand_total            584524 non-null  float64
 7   increment_id           584524 non-null  object 
 8   category_name_1        584360 non-null  object 
 9   sales_commission_code  447346 non-null  object 
 10  discount_amount        584524 non-null  float64
 11  payment_method         584524 non-null  object 
 12  Working Date           584524 non-null  object 
 13  BI Status              584524 non-null  object 
 14   MV                    584524 non-

In [7]:
df.describe()

,item_id,price,qty_ordered,grand_total,discount_amount,Year,Month,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
count,584524.000000,5.845240e+05,584524.000000,5.845240e+05,584524.000000,584524.000000,584524.000000,584513.000000,0.0,0.0,0.0,0.0,0.0
mean,565667.074218,6.348748e+03,1.296388,8.530619e+03,499.492775,2017.044115,7.167654,45790.511965,NaN,NaN,NaN,NaN,NaN
std,200121.173648,1.494927e+04,3.996061,6.132081e+04,1506.943046,0.707355,3.486305,34414.962389,NaN,NaN,NaN,NaN,NaN
min,211131.000000,0.000000e+00,1.000000,-1.594000e+03,-599.500000,2016.000000,1.000000,1.000000,NaN,NaN,NaN,NaN,NaN
25%,395000.750000,3.600000e+02,1.000000,9.450000e+02,0.000000,2017.000000,4.000000,13516.000000,NaN,NaN,NaN,NaN,NaN
50%,568424.500000,8.990000e+02,1.000000,1.960400e+03,0.000000,2017.000000,7.000000,42856.000000,NaN,NaN,NaN,NaN,NaN
75%,739106.250000,4.070000e+03,1.000000,6.999000e+03,160.500000,2018.000000,11.000000,73536.000000,NaN,NaN,NaN,NaN,NaN
max,905208.000000,1.012626e+06,1000.000000,1.788800e+07,90300.000000,2018.000000,12.000000,115326.000000,NaN,NaN,NaN,NaN,NaN


In [8]:
df.head()

,item_id,status,created_at,sku,price,qty_ordered,grand_total,increment_id,category_name_1,sales_commission_code,...,Month,Customer Since,M-Y,FY,Customer ID,Unnamed: 21,Unnamed: 22,Unnamed: 23,Unnamed: 24,Unnamed: 25
0,211131.0,complete,7/1/2016,kreations_YI 06-L,1950.0,1.0,1950.0,100147443,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,1.0,NaN,NaN,NaN,NaN,NaN
1,211133.0,canceled,7/1/2016,kcc_Buy 2 Frey Air Freshener & Get 1 Kasual Bo...,240.0,1.0,240.0,100147444,Beauty & Grooming,\N,...,7.0,2016-7,7-2016,FY17,2.0,NaN,NaN,NaN,NaN,NaN
2,211134.0,canceled,7/1/2016,Ego_UP0017-999-MR0,2450.0,1.0,2450.0,100147445,Women's Fashion,\N,...,7.0,2016-7,7-2016,FY17,3.0,NaN,NaN,NaN,NaN,NaN
3,211135.0,complete,7/1/2016,kcc_krone deal,360.0,1.0,60.0,100147446,Beauty & Grooming,R-FSD-52352,...,7.0,2016-7,7-2016,FY17,4.0,NaN,NaN,NaN,NaN,NaN
4,211136.0,order_refunded,7/1/2016,BK7010400AG,555.0,2.0,1110.0,100147447,Soghaat,\N,...,7.0,2016-7,7-2016,FY17,5.0,NaN,NaN,NaN,NaN,NaN


In [9]:
df.isnull().sum()

item_id                   464051
status                    464066
created_at                464051
sku                       464071
price                     464051
qty_ordered               464051
grand_total               464051
increment_id              464051
category_name_1           464215
sales_commission_code     601229
discount_amount           464051
payment_method            464051
Working Date              464051
BI Status                 464051
 MV                       464051
Year                      464051
Month                     464051
Customer Since            464062
M-Y                       464051
FY                        464051
Customer ID               464062
Unnamed: 21              1048575
Unnamed: 22              1048575
Unnamed: 23              1048575
Unnamed: 24              1048575
Unnamed: 25              1048575
dtype: int64

In [10]:
df.value_counts()

Series([], Name: count, dtype: int64)

In [11]:
df['status'].value_counts()

status
complete          233685
canceled          201249
received           77290
order_refunded     59529
refund              8050
cod                 2859
paid                1159
closed               494
payment_review        57
pending               48
processing            33
holded                31
fraud                 10
pending_paypal         7
exchange               4
\N                     4
Name: count, dtype: int64

In [12]:
df['category_name_1'].value_counts()

category_name_1
Mobiles & Tablets     115710
Men's Fashion          92221
Women's Fashion        59721
Appliances             52413
Superstore             43613
Beauty & Grooming      41496
Soghaat                34011
Others                 29218
Home & Living          26504
Entertainment          26326
Health & Sports        17502
Kids & Baby            16494
Computing              15933
\N                      7850
School & Education      3478
Books                   1870
Name: count, dtype: int64

In [13]:
df['payment_method'].value_counts()

payment_method
cod                  271960
Payaxis               97641
Easypay               82900
jazzwallet            35145
easypay_voucher       31176
bankalfalah           23065
jazzvoucher           15633
Easypay_MA            14028
customercredit         7555
apg                    1758
ublcreditcard           882
cashatdoorstep          732
mcblite                 723
mygateway               669
internetbanking         472
productcredit           125
marketingexpense         45
financesettlement        15
Name: count, dtype: int64

In [14]:
df.duplicated().sum()

np.int64(464050)

In [15]:
df.shape[0] - df.duplicated().sum()

np.int64(584525)

In [16]:
Q1 = df['price'].quantile(0.25)
Q3 = df['price'].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outliers = df[(df['price'] < lower_bound) | (df['price'] > upper_bound)]
print(f"Lower bound: {lower_bound}")
print(f"Upper bound: {upper_bound}")
print(f"Number of outliers in price: {len(outliers)}")

Lower bound: -5205.0
Upper bound: 9635.0
Number of outliers in price: 105687


In [17]:
Q1 = df['grand_total'].quantile(0.25)
Q3 = df['grand_total'].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outliers = df[(df['grand_total'] < lower_bound) | (df['grand_total'] > upper_bound)]
print(f"Lower bound: {lower_bound}")
print(f"Upper bound: {upper_bound}")
print(f"Number of outliers in grand_total: {len(outliers)}")

Lower bound: -8136.0
Upper bound: 16080.0
Number of outliers in grand_total: 77090


In [18]:
Q1 = df['qty_ordered'].quantile(0.25)
Q3 = df['qty_ordered'].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

outliers = df[(df['qty_ordered'] < lower_bound) | (df['qty_ordered'] > upper_bound)]
print(f"Lower bound: {lower_bound}")
print(f"Upper bound: {upper_bound}")
print(f"Number of outliers in qty_ordered: {len(outliers)}")

Lower bound: 1.0
Upper bound: 1.0
Number of outliers in qty_ordered: 79310


In [20]:
# Q1: Which product category has the highest number of orders?
top_category = df['category_name_1'].value_counts().idxmax()
top_category_count = df['category_name_1'].value_counts().max()
print(f"Top category: {top_category} with {top_category_count} orders")

Top category: Mobiles & Tablets with 115710 orders


In [21]:
# Q2: What is the most popular payment method?
top_payment = df['payment_method'].value_counts().idxmax()
top_payment_count = df['payment_method'].value_counts().max()
print(f"Most popular payment method: {top_payment} with {top_payment_count} orders")

Most popular payment method: cod with 271960 orders


In [22]:
# Q3: What percentage of orders were canceled?
canceled = df[df['status'] == 'canceled'].shape[0]
total = df['status'].count()
print(f"Canceled orders: {round((canceled/total)*100, 2)}%")

Canceled orders: 34.43%


In [23]:
# Q4: What is the average order price per year?
print("\nAverage price per year:")
print(df.groupby('Year')['price'].mean().round(2))


Average price per year:
Year
2016.0    3905.65
2017.0    5855.96
2018.0    9295.08
Name: price, dtype: float64


In [24]:
total_revenue = df['grand_total'].sum()
print(f"\nTotal revenue: {round(total_revenue, 2)} PKR")


Total revenue: 4986351289.57 PKR


In [25]:
top_month = df['Month'].value_counts().idxmax()
top_month_count = df['Month'].value_counts().max()
print(f"\nBusiest month: {top_month} with {top_month_count} orders")


Busiest month: 11.0 with 155456 orders


In [26]:
discounted = df[df['discount_amount'] > 0].shape[0]
print(f"\nOrders with discount: {round((discounted/total)*100, 2)}%")


Orders with discount: 35.62%


In [27]:
print("\nOrders per year:")
print(df.groupby('Year')['item_id'].count())


Orders per year:
Year
2016.0    133909
2017.0    290920
2018.0    159695
Name: item_id, dtype: int64


# Data Quality Report
## Pakistan's Largest E-Commerce Dataset

### 1. Dataset Overview
This dataset contains e-commerce transaction data from Pakistan's largest 
online shopping platform. It covers orders from March 2016 to August 2018. 
The dataset was sourced from Kaggle 
(https://www.kaggle.com/datasets/zusmani/pakistans-largest-ecommerce-dataset) 
and was collected by Zeeshan Usmani. After loading, the dataset contains 
1,048,575 rows and 26 columns, however only 584,524 rows contain real data.

### 2. Missing Values
Several columns contain significant missing values. The most critical is 
sales_commission_code with 601,229 missing values — more than the total real 
rows in the dataset. category_name_1 has 464,215 missing values. Additionally, 
columns Unnamed: 21 through Unnamed: 25 have zero non-null values meaning they 
are completely empty and serve no purpose in analysis.

### 3. Dead Rows and Duplicate Rows
Out of 1,048,575 total rows loaded, 464,051 rows are completely empty. These 
blank rows were loaded from the bottom of the CSV file. This means 44.2% of 
the loaded data is unusable. The duplicated().sum() function confirmed 464,050 
duplicate rows, which matches the empty row count exactly.

### 4. Outliers
Using the IQR method, significant outliers were found in three numeric columns.
The price column has 105,687 outliers with an upper bound of 9,635 PKR. The 
grand_total column has 77,090 outliers with an upper bound of 16,080 PKR. The 
qty_ordered column has 79,310 outliers — this happened because 75% of orders 
contain exactly 1 item, making the IQR equal to zero and flagging anything 
above 1 as an outlier.

### 5. Suspicious and Incorrect Values
Multiple data quality issues were found beyond missing values. The grand_total 
column contains negative values with a minimum of -1,594 PKR which is 
impossible for an order total. The discount_amount column also contains 
negative values with a minimum of -599 PKR. The status and category_name_1 
columns contain '\N' as text values instead of proper null values — this is a 
database export error. The Year and Month columns are stored as float64 instead 
of integers which is incorrect. Additionally 34.43% of all orders were canceled 
which may indicate customer experience problems.

### 6. Week 2 Plan
In Week 2 the following issues will be addressed. First the 464,051 empty rows 
will be dropped using dropna(). The five completely empty unnamed columns will 
be deleted. The '\N' text values in status and category_name_1 will be replaced 
with proper NaN values using replace(). Negative values in grand_total and 
discount_amount will be investigated and either corrected or removed. The Year 
and Month columns will be converted from float64 to integer using astype(). 
Finally outliers in price and grand_total will be capped at their respective 
upper bounds to prevent them from skewing future analysis.